### ISSUE TITLE
Problem: breaking changes occur due to lack of tests

### ISSUE THREAD

**Issue by samuelralak** [Blockrazor/blockrazor#636]
Problem: there are currently only 16 tests and no way of knowing whether or not contrubutors introduce breaking changes; an example would be in this [closed issue](https://github.com/Blockrazor/blockrazor/issues/619) where @gazhayes says the feature was initially working but was somehow broken. introducing more tests would help avoid such problems.

Possible solution: Write more tests for server methods and possibly the client UI; we can definitely find more things to test as stated [here](https://github.com/Blockrazor/blockrazor/issues/368). I propose that this process should be done piecemeal; a corresponding issue created for each folder that needs testing.

This process would also help new contributors familiarise themselves with the codebase quick and possibly improve their understanding on how things work.

@gazhayes what are your thoughts on this?

ping @gazhayes @AshHimself @MadanBhandari @janat08 @anbud @sjmackenzie

**Comment by gazhayes** [comment]
Most new code after #303 should have tests written. 

We should definitely be using tests to stop things from breaking, but I agree it needs to be done piecemeal.

How about you choose a module, log an issue for it, and send a PR with some tests to solve it. Then @anbud can review and see if anything should be done differently before proceeding with more of them.

**Comment by janat08** [comment]
you're vying for the 100$ reward? this isn't actionable.

**Comment by sjmackenzie** [comment]
The most recent comment is a sign the reward system is having the reverse effect. Probably best to carefully monitor the effects and move quickly to deprecate it. But it's too early to tell atm.

**Comment by gazhayes** [comment]
@janat08 is right in many ways, this is more of a low-priority issue and not an important problem right now. But it's still worth raising because it *is* a problem.

We don't need a comprehensive set of tests, but we do need them for things that we really don't want to break and are hard to detect when they break. This does happen, there have been a few times now where functionality has been broken and I don't find out till I stumble on it a week later.

On second thoughts though @samuelralak it's probably better to just mark this as low priority and move on to other things, but keep testing in mind whenever working on something.

**Comment by samuelralak** [comment]
@gazhayes thats alright, I understand.<br>
When I created this issue I had no reward in mind nor was I trying to vie for anything; so I honestly do not understand what this means `"you're vying for the 100$ reward? this isn't actionable."` or where its coming from. <br>
One of my motivations for bringing this up is the challenge in coherently understanding the code for new comers like me and ofcourse the risk of introducing breaking changes. I thought it would be great if we 1. make things easier for future contributors 2. avoid shooting ourselves in the foot when the product if officially published.<br>
I will definitely keep testing in mind when working on future changes.

**Comment by gazhayes** [comment]
> When I created this issue I had no reward in mind nor was I trying to vie for anything; so I honestly do not understand what this means "you're vying for the 100$ reward? this isn't actionable." or where its coming from. 

That's just @janat08 making a joke, don't worry ;)

**Comment by janat08** [comment]
No, I'm truly suspicious and cynical.

Anyway, with the developer onboarding in mind, jest's snapshot testing will just give you the diff of before and after of whatever run, so it's much more succinct, and comprehensive for it's ease of set up and maintenance (as in you can just ignore it failing altogether). I'd like to run an example pull request.

**Comment by gazhayes** [comment]
> jest's snapshot testing
This is simply UI rendering comparisons, right? Not checking whether user actions such as commenting etc still work?

I don't really see the benefit of this or the problem it's solving. So far in the history of Blockrazor nothing in the UI 'image' has been broken. User actions have been broken, and backend stuff has been broken, but not the UI. 

Also the UI is in constant flux, so the test would almost never pass. Everyone will just be in the habit of ignoring it. I also don't want to discourage new people from making improvements in the UI because it breaks a test. 

Finally, you can't do things like merge PRs on a phone when you have failing tests, this would annoy the hell out of me.

It could be more useful once the UI is refined and we aren't making changes to it every day, but I don't know about now.

How long does it take you to set it up and do an example?

**Comment by janat08** [comment]
Theoretically, you run a template save the code and run an event, and save the difference between old save and new save, then in next tests the diff is compared against that. Works the same way with back-end testing, save DB code, run method, save the diff, and test against that.

I just realized that it won't be as simple with blaze, although there appear to be packages that could help with just rendering a template to a file. I'm newish to testing and snapshots so it might take a couple of hours to set up, but it's only isolating the code to test and running it which gives you a "thorough test".

**Comment by gazhayes** [comment]
OK. There are more critical issues at the moment, but we can definitely use snapshot testing later when the UI is stable, it will start solving problems then.
